packages/fdc3-agent-proxy/src/listeners/AbstractListener.ts

```diff
@@ -1,54 +1,115 @@
+  | PrivateChannelUnsubscribeEventListenerResponse;
+
 /**
  * Common to all listeners - they need to be registered and unregistered with messaging and also
- * send notification messsages when connected and disconnected
+ * send notification messages when connected and disconnected
  */
-export abstract class AbstractListener<X> implements RegisterableListener {
+export abstract class AbstractListener<X, Y extends SubscriptionRequest> implements RegisterableListener {
   readonly messaging: Messaging;
-  readonly handler: X;
+  private readonly subscribeRequestType: Y['type'];
+  private readonly subscribeResponseType: SubscriptionResponse['type'];
+  private readonly unsubscribeRequestType: UnsubscribeRequest['type'];
+  private readonly unsubscribeResponseType: UnsubscribeResponse['type'];
+  private readonly subscriptionPayload: Y['payload'];
   id: string | null = null;
+  readonly handler: X;
 
   constructor(
     messaging: Messaging,
-    private readonly payloadDetails: Record<string, string | null>,
+    subscriptionPayload: Y['payload'],
     handler: X,
-    private readonly subscribeType: string,
-    private readonly unsubscribeType: string
+    subscribeRequestType: Y['type'],
+    subscribeResponseType: SubscriptionResponse['type'],
+    unsubscribeRequestType: UnsubscribeRequest['type'],
+    unsubscribeResponseType: UnsubscribeResponse['type']
   ) {
     this.messaging = messaging;
     this.handler = handler;
+    this.subscriptionPayload = subscriptionPayload;
+    this.subscribeRequestType = subscribeRequestType;
+    this.subscribeResponseType = subscribeResponseType;
+    this.unsubscribeRequestType = unsubscribeRequestType;
+    this.unsubscribeResponseType = unsubscribeResponseType;
   }
 
-  abstract filter(m: any): boolean;
+  abstract filter(m: AgentEventMessage): boolean;
 
-  abstract action(m: any): void;
+  abstract action(m: AgentEventMessage): void;
+
+  async unsubscribe(): Promise<void> {
+    if (this.id) {
+      this.messaging.unregister(this.id!);
+      const unsubscribeMessage: UnsubscribeRequest = {
+        meta: this.messaging.createMeta(),
+        payload: {
+          listenerUUID: this.id,
+        },
+        type: this.unsubscribeRequestType,
+      };
+      await this.messaging.exchange<UnsubscribeResponse>(unsubscribeMessage, this.unsubscribeResponseType);
+      return;
+    } else {
+      console.error("This listener doesn't have an id and hence can't be removed!");
+    }
+  }
 
-  async listenerNotification(type: string): Promise<string | null> {
-    const requestType = `${type}Request`;
-    const responseType = `${type}Response`;
-    const notificationMessage = {
+  async register(): Promise<void> {
+    const subscribeMessage: Y = {
       meta: this.messaging.createMeta(),
-      payload: this.id ? { listenerUUID: this.id } : { ...this.payloadDetails },
-      type: requestType,
-    };
+      payload: this.subscriptionPayload,
+      type: this.subscribeRequestType,
+    } as Y;
```

Comment:
Is the `as Y` necessary?

Reply:
Apparently so - I don't understand why. The error message is: 

> 'Y' could be instantiated with an arbitrary type which could be unrelated to '{ meta: Y["meta"]; payload: Y["payload"]; type: Y["type"]; }'

This happens even if I type the individual properties: meta, payload, type):

```ts
const subscribeMessage: Y = {
      meta: this.messaging.createMeta() as Y['meta'],
      payload: this.subscriptionPayload as Y['payload'],
      type: this.subscribeRequestType as Y['type'],
    };
```

So there clearly something about TypeScript that I don't understand here. If you can figure it out @julianna-ciq you'd be teaching me something ;-)